### PR TITLE
Constrain constant values to a single location in code

### PR DESCRIPTION
**Pull request by luke-jr**

### CODE REVIEW

**Comment by paveljanik** [comment]
No wallet build fails with:

  CXX    libbitcoin_server_a-init.o
init.cpp: In function ‘std::string HelpMessage(HelpMessageMode)’:
init.cpp:363:156: error: ‘DEFAULT_WALLET_DBLOGSIZE’ was not declared in this scope
init.cpp:368:121: error: ‘DEFAULT_FLUSHWALLET’ was not declared in this scope
init.cpp:394:128: error: ‘DEFAULT_WALLET_PRIVDB’ was not declared in this scope

**Comment by jgarzik** [comment]
concept ACK.

* seemed like the REST constant should go in a rest, not rpc, file

**Comment by petertodd** [comment]
concept ACK

**Comment by luke-jr** [comment]
@jgarzik I put it in rpcserver.h because the -rest option is used in rpcserver.cpp, which handles all HTTP requests. There is no rest.h - do you really want me to make one just for this?

**Comment by luke-jr** [comment]
@paveljanik Fixed.

File: src/main.h [hunk 1]
```diff
@@ -90,6 +90,13 @@ static const unsigned int DATABASE_WRITE_INTERVAL = 60 * 60;
 static const unsigned int DATABASE_FLUSH_INTERVAL = 24 * 60 * 60;
 /** Maximum length of reject messages. */
 static const unsigned int MAX_REJECT_MESSAGE_LENGTH = 111;
+static const unsigned int DEFAULT_LIMITFREERELAY = 15;
+static const bool DEFAULT_RELAYPRIORITY = true;
+
+static const bool TXINDEX_DEFAULT = false;
```
Comment (Diapolo) [inline src/main.h]: Why did you call this TXINDEX_DEFAULT instead of DEFAULT_TXINDEX?

File: src/script/standard.cpp [hunk 2]
```diff
@@ -53,7 +54,7 @@ bool Solver(const CScript& scriptPubKey, txnouttype& typeRet, vector<vector<unsi
         mTemplates.insert(make_pair(TX_MULTISIG, CScript() << OP_SMALLINTEGER << OP_PUBKEYS << OP_SMALLINTEGER << OP_CHECKMULTISIG));
 
         // Empty, provably prunable, data-carrying output
-        if (GetBoolArg("-datacarrier", true))
```
Comment (Diapolo) [inline src/script/standard.cpp]: Why is this removed and now defaults just to true?

File: src/main.h (hunk 1 above)
Comment (luke-jr) [inline src/main.h]: Dunno, changed for consistency.

File: src/util.h [hunk 3]
```diff
@@ -48,6 +48,13 @@ extern bool fLogIPs;
 extern volatile bool fReopenDebugLog;
 extern CTranslationInterface translationInterface;
 
+static const char BITCOIN_CONF_FILENAME[] = "bitcoin.conf";
```
Comment (laanwj) [inline src/util.h]: Declaring the constant and its value in a header won't work well for data structures that require a larger amount of memory or indirected pointers, such as strings, structs, objects etc. In worst case this causes errors in some compilers, in best case this causes unnecessary memory use.

**Comment by laanwj** [comment]
Concept ACK

File: src/util.h (hunk 3 above)
Comment (laanwj) [inline src/util.h]: So please move the content of the variable and `BITCOIN_PID_FILENAME` to util.cpp
Apart from that I'd like to merge this

**Comment by laanwj** [comment]
Needs rebase + string nit addresed

**Comment by dcousens** [comment]
concept ACK

**Comment by jgarzik** [comment]
concept ACK - as @laanwj says it appears merge-ready once rebased + nits addressed

**Comment by luke-jr** [comment]
Rebased, addressed nit, added equivalent commit for -uiplatform. (Travis passes.)

File: src/init.cpp [hunk 4]
```diff
@@ -476,15 +487,15 @@ std::string HelpMessage(HelpMessageMode mode)
...
-        strUsage += HelpMessageOpt("-choosedatadir", _("Choose data directory on startup (default: 0)"));
+        strUsage += HelpMessageOpt("-choosedatadir", strprintf(_("Choose data directory on startup (default: %u)"), DEFAULT_CHOOSE_DATADIR));
         strUsage += HelpMessageOpt("-lang=<lang>", _("Set language, for example \"de_DE\" (default: system locale)"));
         strUsage += HelpMessageOpt("-min", _("Start minimized"));
         strUsage += HelpMessageOpt("-rootcertificates=<file>", _("Set SSL root certificates for payment request (default: -system-)"));
-        strUsage += HelpMessageOpt("-splash", _("Show splash screen on startup (default: 1)"));
-        if (showDebug) {
-            strUsage += HelpMessageOpt("-uiplatform", "Select platform to customize UI for (one of windows, macosx, other; default: platform compiled on)");
+        strUsage += HelpMessageOpt("-splash", strprintf(_("Show splash screen on startup (default: %u)"), DEFAULT_SPLASHSCREEN));
+        if (showDebug && !uiInterface.DefaultUIPlatform.empty()) {
```
Comment (laanwj) [inline src/init.cpp]: I don't think we should use `uiInterface` here, nor put this uiplatform id as a variable on the UI interface (as bitcoind has no business with it). I'd just check for `mode == HMM_BITCOIN_QT`, like for the other options.

Comment (laanwj) [inline src/init.cpp]: Hmm I understand now, you want to show the actual default here. Ok then it makes sense.
We should *never* have moved the UI-specific options back to init, but that's  not your fault.

**Comment by dcousens** [comment]
utACK,  however I didn't cross verify the constants.

**Comment by MarcoFalke** [comment]
utACK 29266b64e8d17baee18d42a93a8f1a790ad858ca.

**Comment by MarcoFalke** [comment]
Does this need more reviewers or can it be merged as is? (Any further merge conflicts would only retrigger review, so let's try to finish with this PR.)

**Comment by MarcoFalke** [comment]
Needs rebase caused by #6235.

**Comment by dcousens** [comment]
@luke-jr this needs rebase
@laanwj anything holding back from merge?

**Comment by sipa** [comment]
Fixed by #6961.

**Comment by MarcoFalke** [comment]
@luke-jr are you going to rebase (3/3) 29266b6?
